### PR TITLE
overlord/snapstate: perform a single reboot when updating boot base and kernel

### PR DESCRIPTION
**Pull request by bboozzoo**
The branch changes snapd to perform only a single reboot when updating both kernel and the boot base.

Note that the branch is still missing a managers test, as I failed to make settle() behave nicely and not loop until hits the timeout.

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -1200,6 +1200,92 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, userID int
...
+			// only kernel in this update
+			linkSnap, _ = findTaskOfKind(kernelTs, "link-snap")
+		} else {
+			// only boot base in this update
+			linkSnap, _ = findTaskOfKind(bootBaseTs, "link-snap")
+		}
+		if linkSnap != nil {
+			// we're only updating the base or the kernel, so the link-snap
+			// task can request a reboot
+			linkSnap.Set("can-reboot", true)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1230]: isn't it simpler to keep  can-reboot missing to mean reboot, so we don't have to do this? and we have to set can-reboot only if we have more than rebooting task?

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1396,12 +1482,17 @@ func doUpdate(ctx context.Context, st *state.State, names []string, updates []mi
...
 		}
-		// keep track of kernel/gadget udpates
+		// keep track of kernel/gadget updates
 		switch update.Type() {
 		case snap.TypeKernel:
 			kernelTs = ts
 		case snap.TypeGadget:
 			gadgetTs = ts
+		case snap.TypeBase, snap.TypeOS:
+			if update.InstanceName() == deviceCtx.Model().Base() {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1492]: does this work for core? isn't base "" then?

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -1200,6 +1200,92 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, userID int
 // consider.
 type updateFilter func(*snap.Info, *SnapState) bool
 
+func rearrangeBaseKernelForSingleReboot(kernelTs, bootBaseTs *state.TaskSet) error {
+	var findTaskOfKind = func(ts *state.TaskSet, kind string) (*state.Task, int) {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:1204]: we could have doInstall use TaskSet.MarkEdge more and use MaybeEdge  here instead, and instead of the indexes we would mark both the link-snap and the before/after tasks

File: overlord/snapstate/handlers.go [hunk 4]
```diff
@@ -1564,7 +1564,23 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:1564]: do we reboot again along the undo path? do we need to do something about that? I suppose at that point we will have reboot more than once anyway

**Review by pedronis — COMMENTED** [review]
direction looks good,  a question some comments/suggestions to simplify some things or follow preexisting patterns

File: overlord/snapstate/handlers.go (hunk 4 above)
Comment (bboozzoo) [reply overlord/snapstate/handlers.go:1564]: Yes, there would be restart requests for the base and for the kernel.

File: overlord/snapstate/snapstate.go (hunk 2 above)
Comment (bboozzoo) [reply overlord/snapstate/snapstate.go:1480]: There's a bit of complication with core, so I've added a comment below. When core is part of the update, we implicitly add its tasks as dependency for all tasks from other snaps. In case of a kernel (using core base) and core update, we end up with kernel tasks waiting for all core tasks. Thus introducing additional wait in core's auto-connect on link-snap of the kernel creates a wait loop. I think that technically it would be possible to tweak the wait lists such that we can reorder the tasks to achieve a single reboot setup, but I left it out for now.

File: overlord/snapstate/handlers.go [hunk 5]
```diff
@@ -1570,7 +1570,24 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
...
+		}
+		if !cannotReboot {
+			// either the task was created before that variable was
+			// introduced or the task can request a reboot
+			canReboot = true
+		} else {
+			t.Logf("reboot postponed to later tasks")
+		}
+	}
+	if !reboot || canReboot {
```
Comment (mardy) [inline overlord/snapstate/handlers.go:1588]: Shouldn't this be `if reboot || canReboot {`?

I know that you didn't introduce the `reboot` variable with this change, but IMHO this branch would be a good occasion to rename it (`mustReboot`, `reboot required`, `willReboot`, `didReboot`,etc., depending on what the variable actually means).

File: overlord/snapstate/snapstate.go [hunk 6]
```diff
@@ -1200,6 +1200,80 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, userID int
...
+	//
+	// where (r) denotes the task that can effectively request a reboot
+
+	linkSnapKernel, lsKernelIdx := findTaskOfKind(kernelTs, "link-snap")
+	autoConnectKernel, _ := findTaskOfKind(kernelTs, "auto-connect")
+	if linkSnapKernel == nil || autoConnectKernel == nil {
+		return fmt.Errorf("internal error: cannot identify link-snap or auto-connect for the kernel snap")
+	}
+	kernelLanes := linkSnapKernel.Lanes()
+	linkSnapBase, _ := findTaskOfKind(bootBaseTs, "link-snap")
```
Comment (mardy) [inline overlord/snapstate/snapstate.go:1242]: nitpick^100000: add an empty line between these two blocks

File: overlord/snapstate/snapstate.go [hunk 7]
```diff
@@ -1200,6 +1200,80 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, userID int
...
+
+	for _, lane := range kernelLanes {
+		linkSnapBase.JoinLane(lane)
+		autoConnectBase.JoinLane(lane)
+	}
+	for _, lane := range baseLanes {
+		linkSnapKernel.JoinLane(lane)
+		autoConnectKernel.JoinLane(lane)
+	}
+	// make link-snap base wait for the last pre-link-snap-kernel task
```
Comment (mardy) [inline overlord/snapstate/snapstate.go:1257]: I went and grepped the source code for a "pre-link-*" task kind, but I found nothing. If we don't have a specific "pre-link" type, I recommend rewriting this comment to something like "_...for the last task before the kernel's "link-snap"_".

It's still not clear why we need to wait for that, though: if the `snap-link(base)` task is the first one we care for (according to the graph from the comment above), it would seem it can be started even earlier.

File: overlord/snapstate/snapstate.go [hunk 8]
```diff
@@ -1200,6 +1200,80 @@ func UpdateMany(ctx context.Context, st *state.State, names []string, userID int
...
+	linkSnapBase.WaitFor(kernelTs.Tasks()[lsKernelIdx-1])
+	// order: link-snap-base -> link-snap-kernel
+	linkSnapKernel.WaitFor(linkSnapBase)
+	// order: link-snap-kernel -> auto-connect-base
+	autoConnectBase.WaitFor(linkSnapKernel)
+	// order: auto-connect-base -> auto-connect-kernel
+	autoConnectKernel.WaitFor(autoConnectBase)
+	// make the first post auto-connect base wait for auto-connect kernel,
```
Comment (mardy) [inline overlord/snapstate/snapstate.go:1265]:
```suggestion
	// make the first task after auto-connect-base wait for auto-connect-kernel,
```

**Review by mardy — COMMENTED** [review]
Questions and nitpicks :-)

File: overlord/snapstate/handlers.go (hunk 5 above)
Comment (bboozzoo) [reply overlord/snapstate/handlers.go:1588]: Hmm `needsReboot`?

As for the if check, I see how it can be confusing. I'll tweak this part to make it more clear, but yeah the condition is `!reboot || canReboot` (canReboot is only true when we need a reboot and it's allowed, and !reboot is true when there may be a need to daemon restart only)

Comment (bboozzoo) [reply overlord/snapstate/handlers.go:1588]: renamed to needsReboot

File: overlord/snapstate/snapstate.go (hunk 6 above)
Comment (bboozzoo) [reply overlord/snapstate/snapstate.go:1242]: added

File: overlord/snapstate/snapstate.go (hunk 2 above)
Comment (bboozzoo) [reply overlord/snapstate/snapstate.go:1480]: As discussed with @pedronis we will limit the feature to UC18+ devices.

File: overlord/snapstate/snapstate.go (hunk 7 above)
Comment (bboozzoo) [reply overlord/snapstate/snapstate.go:1257]: I've tweaked the comment. Ordering is just to ensure that all tasks before link-snap of the kernel have finished before we ru link-snap of the base.

**Review by mardy — APPROVED** [review]
LGTM, thanks for the updates!

File: overlord/snapstate/snapstate_update_test.go [hunk 9]
```diff
@@ -7073,15 +7083,145 @@ func (s *snapmgrTestSuite) TestUpdateBaseKernelSingleRebootHappy(c *C) {
...
+		"core18/11", "kernel-core18/11",
+		"auto-connect:Doing-core18/11", "auto-connect:Doing-kernel-core18/11",
 	})
 	c.Assert(ops[6:], testutil.DeepUnsortedMatches, []string{
-		"cleanup-trash-core18", "cleanup-trash-kernel",
+		"cleanup-trash-core18", "cleanup-trash-kernel-core18",
+	})
+}
+
+func (s *snapmgrTestSuite) TestUpdateBaseKernelSingleRebootNotWithCoreHappy(c *C) {
```
Comment (pedronis) [inline overlord/snapstate/snapstate_update_test.go:7200]: the name is a bit hard to parse,  s/Not/Impossible/ s/Not/Unsupported/ ?

File: overlord/snapstate/snapstate_update_test.go [hunk 10]
```diff
@@ -7001,3 +7008,457 @@ func findLastTask(chg *state.Change, kind string) *state.Task {
...
+	}))
+
+	restore = snapstatetest.MockDeviceModel(MakeModel(map[string]interface{}{
+		"kernel": "kernel-core18",
+		"base":   "core18",
+	}))
+	defer restore()
+
+	siKernel := snap.SideInfo{
+		RealName: "kernel-core18",
```
Comment (pedronis) [inline overlord/snapstate/snapstate_update_test.go:7033]: pc-kernel has its base unset,  should we have/keep also a test which such base-unset kernel?

**Review by pedronis — APPROVED** [review]
looks good, one question about the changed test vs old test

**Comment by anonymouse64** [comment]
Unfortunately this PR also doesn't cherry-pick cleanly, nor was it squash merged, @bboozzoo would you mind opening a backport PR to `release/2.53` for this? Thanks
